### PR TITLE
add_subdomain_names

### LINKED ISSUE #27474: Allow specifying subdomain names that do not yet exist in the mesh
### Discussed in https://github.com/idaholab/moose/discussions/27415

<div type='discussions-op-text'>

<sup>Originally posted by **grunerjmeier** April 19, 2024</sup>
Dear Community,

Following the procedure recommended [here](https://mooseframework.inl.gov/framework/contributing.html), I would like to propose the property "add_subdomain_names" for the object "MooseMesh" for discussion.

**Motivation and current situation**
Using the property `add_subdomain_ids` of the [MooseMesh base class](https://mooseframework.inl.gov/source/mesh/MooseMesh.html) a user can add (initially empty) subdomains to an existing mesh for later usage. Up to my knowledge today only subdomain IDs can be added and no name is provided for these subdomains. To ease reading and debugging of the Moose input file it would be great to also provide the possibility to add subdomain names. For this purpose I like to propose an additional property for the object "MooseMesh": add_subdomain_names

**Proposed property "add_subdomain_names" for the object MooseMesh**
Following the design pattern of existing objects and their properties, e.g. with "symbol_names" and "symbol_values" in [ParsedFunction](https://mooseframework.inl.gov/source/functions/MooseParsedFunction.html) I like to add the property `add_subdomain_names` as a companion property to the existing property `add_subdomain_ids`. While `add_subdomain_ids` is taking a vector of subdomain ids, the additional (optional) `add_subdomain_names` provides the corresponding vector of SubdomainNames. The property add_subdomain_names would only be valid, if add_subdomain_ids is defined.

A input file using this property might look like:
```
[Mesh]
  ...
  add_subdomain_ids  =  '5      6      7'
  add_subdomain_names = 'BlockX BlockY BlockZ'
  ...
[]
```

**Impact to existing Moose apps**
As this proposal only adds functionality. Existing Moose apps should be able to cope with the proposed changes without any problems.

Please find a crude proof of concept [here](https://github.com/grunerjmeier/moose/blob/279ee92f19f34c3c9ebed262898e748d25af1414/framework/src/mesh/MooseMesh.C#L169) and [there](https://github.com/grunerjmeier/moose/blob/279ee92f19f34c3c9ebed262898e748d25af1414/framework/src/mesh/MooseMesh.C#L409).

If this proposal meets with interest, I need support in preparing the code before I can create a PR. The proof-of-concept works for me, but does not include support for "parallel mesh", for example.

I would be delighted if you could consider this proposal.
</div>

### PR DESCRIPTION
**Pull request by jmeier**
refs #27415

## Reason
Using the property add_subdomain_ids of the MooseMesh base class a user can add (initially empty) subdomains to an existing mesh for later usage. Up to my knowledge today only subdomain IDs can be added and no name is provided for these subdomains. This PR proposes the property add_subdomain_names addessing this issue.

## Design
This PR adds the proeprty add_subdomain_names to the MooseMesh base class

## Impact
This PR only adds functionality. Existing Moose apps should be able to cope with the proposed changes without any problems.

### CODE REVIEW

**Comment by grunerjmeier** [comment]
As encouraged by @GiudGiud, I would like to point out two questions that are unclear in the code of this PR:

- The code calls `setSubdomainName(...)` within `MooseMesh::prepare`. Is this permitted?
- `_communicator.set_union(_mesh_subdomains);` publishes only the added subdomain ids (not the names). Do we have to publish the subdomain names too? If yes, how?

**Comment by GiudGiud** [comment]
> The code calls setSubdomainName(...) within MooseMesh::prepare. Is this permitted?

this is ok afaict

> _communicator.set_union(_mesh_subdomains); publishes only the added subdomain ids (not the names). Do we have to publish the subdomain names too? If yes, how?

every MPI process has access to the input file and runs this code. so you dont need to communicate the name-id mapping

I think the way you have it now works, you dont need to enable "name-only" mode for now

File: framework/src/mesh/MooseMesh.C [hunk 1]
```diff
@@ -401,6 +406,26 @@ MooseMesh::prepare(const MeshBase * const mesh_to_clone)
...
     const auto add_subdomain_id = getParam<std::vector<SubdomainID>>("add_subdomain_ids");
     _mesh_subdomains.insert(add_subdomain_id.begin(), add_subdomain_id.end());
+
+    // do we have subdomain names explicitly given?
+    if (isParamValid("add_subdomain_names"))
+    {
+      const auto add_subdomain_name = getParam<std::vector<SubdomainName>>("add_subdomain_names");
+      const int n_ids = add_subdomain_id.size();
+      const int n_names = add_subdomain_name.size();
+
```
Comment (dschwen) [inline framework/src/mesh/MooseMesh.C:416]: Use `getParam<SubdomainID, SubdomainName>("add_subdomain_ids", "add_subdomain_names")` to get a vector of pairs with automatic checking for equal number of entries.

Comment (grunerjmeier) [reply framework/src/mesh/MooseMesh.C:416]: Dear @dschwen, This looks elegant! 

But then the user *must* provide both, `add_subdomain_ids` and `add_subdomain_names`? Wouldn't that break all existing models that only provide `add_subdomain_ids` to date (and not `add_subdomain_names`)?

Or to put it another way: Will Moose demand in future that both `add_subdomain_names` and `add_subdomain_names` must be given?

File: framework/src/mesh/MooseMesh.C [hunk 2]
```diff
@@ -401,6 +406,26 @@ MooseMesh::prepare(const MeshBase * const mesh_to_clone)
   {
     const auto add_subdomain_id = getParam<std::vector<SubdomainID>>("add_subdomain_ids");
     _mesh_subdomains.insert(add_subdomain_id.begin(), add_subdomain_id.end());
+
+    // do we have subdomain names explicitly given?
+    if (isParamValid("add_subdomain_names"))
+    {
+      const auto add_subdomain_name = getParam<std::vector<SubdomainName>>("add_subdomain_names");
```
Comment (lindsayad) [inline framework/src/mesh/MooseMesh.C:413]:
```suggestion
      const auto & add_subdomain_name = getParam<std::vector<SubdomainName>>("add_subdomain_names");
```

File: framework/src/mesh/MooseMesh.C [hunk 3]
```diff
@@ -401,6 +406,26 @@ MooseMesh::prepare(const MeshBase * const mesh_to_clone)
...
+      const int n_ids = add_subdomain_id.size();
+      const int n_names = add_subdomain_name.size();
+
+      if (n_ids != n_names)
+        mooseError("Number of items provided in add_subdomain_ids and add_subdomain_names does not "
+                   "match.");
+
+      for (int i = 0; i < n_ids; ++i)
```
Comment (lindsayad) [inline framework/src/mesh/MooseMesh.C:421]:
```suggestion
      for (const auto i : make_range(n_ids))
```

File: framework/src/mesh/MooseMesh.C [hunk 4]
```diff
@@ -401,6 +406,26 @@ MooseMesh::prepare(const MeshBase * const mesh_to_clone)
...
+        setSubdomainName(add_subdomain_id[i], add_subdomain_name[i]);
+    };
+  }
+  else if (isParamValid("add_subdomain_names"))
+  {
+    // the user has defined add_subdomain_names, but not add_subdomain_ids
+    mooseError("In combination with add_subdomain_names, add_subdomain_ids must be defined.");
   }
```
Comment (lindsayad) [inline framework/src/mesh/MooseMesh.C:429]:
```suggestion
    // the user has defined add_subdomain_names, but not add_subdomain_ids
    mooseError("In combination with add_subdomain_names, add_subdomain_ids must be defined.");
```

**Review by lindsayad — CHANGES_REQUESTED** [review]

File: framework/src/mesh/MooseMesh.C (hunk 1 above)
Comment (dschwen) [reply framework/src/mesh/MooseMesh.C:416]: Not if you keep the `isParamValid("add_subdomain_names")` and conditionally fetch the vector of pairs

File: framework/src/mesh/MooseMesh.C [hunk 5]
```diff
@@ -397,11 +402,26 @@ MooseMesh::prepare(const MeshBase * const mesh_to_clone)
     _mesh_subdomains.insert(elem->subdomain_id());
 
   // add explicitly requested subdomains
-  if (isParamValid("add_subdomain_ids"))
+  if (isParamValid("add_subdomain_ids") && !isParamValid("add_subdomain_names"))
   {
+    // only subdomain ids are explicitly given
     const auto add_subdomain_id = getParam<std::vector<SubdomainID>>("add_subdomain_ids");
```
Comment (lindsayad) [inline framework/src/mesh/MooseMesh.C:408]:
```suggestion
    const auto & add_subdomain_id = getParam<std::vector<SubdomainID>>("add_subdomain_ids");
```

File: framework/src/mesh/MooseMesh.C [hunk 6]
```diff
@@ -397,11 +402,26 @@ MooseMesh::prepare(const MeshBase * const mesh_to_clone)
...
     _mesh_subdomains.insert(add_subdomain_id.begin(), add_subdomain_id.end());
+  } 
+  else if (isParamValid("add_subdomain_ids") && isParamValid("add_subdomain_names")) 
+  {
+    const auto add_subdomain = getParam<SubdomainID, SubdomainName>("add_subdomain_ids", "add_subdomain_names");
+    for (const auto & i : add_subdomain)
```
Comment (lindsayad) [inline framework/src/mesh/MooseMesh.C:414]:
```suggestion
    for (const auto & [sub_id, sub_name] : add_subdomain)
```

File: framework/src/mesh/MooseMesh.C [hunk 7]
```diff
@@ -397,11 +402,26 @@ MooseMesh::prepare(const MeshBase * const mesh_to_clone)
...
+    {
+      // add subdomain id
+      _mesh_subdomains.insert(i.first);
```
Comment (lindsayad) [inline framework/src/mesh/MooseMesh.C:417]:
```suggestion
      _mesh_subdomains.insert(sub_id);
```

File: framework/src/mesh/MooseMesh.C [hunk 8]
```diff
@@ -397,11 +402,26 @@ MooseMesh::prepare(const MeshBase * const mesh_to_clone)
...
+      // set name of the subdomain just added
+      setSubdomainName(i.first, i.second);
```
Comment (lindsayad) [inline framework/src/mesh/MooseMesh.C:419]:
```suggestion
      setSubdomainName(sub_id, sub_name);
```

**Review by lindsayad — CHANGES_REQUESTED** [review]
almost there, appreciate the contribution!

**Comment by grunerjmeier** [comment]
Dear @lindsayad, Thank you very much for your support.

**Comment by lindsayad** [comment]
Ugh, we need to stop making tests that are predicated on our syntax never changing. @brandonlangley it looks like you created the failing MOOSE server test which perhaps is relying on the number of `MooseMesh` parameters to be exactly equal to 47? My quick perusal and conclusion could be wrong

**Comment by grunerjmeier** [comment]
GitHub tells me, I have to address one more change request [here](https://github.com/idaholab/moose/pull/27420#pullrequestreview-2018337918). I do not know how to resolve this. Can you hint me?

**Comment by lindsayad** [comment]
There's nothing you can do about that. That is for me to "fix". Once all my requested changes have been addressed, then I will approve and that label will get removed. Do you know how to make a test? A test would be good to verify that your code is working

**Comment by grunerjmeier** [comment]
> Do you know how to make a test?

Unfortunately, no. Can we use the example [here](https://github.com/idaholab/moose/discussions/26941) as a test (adding add_subdomain_names for sure)? 
What do I have to do to make a test from #26941?

**Comment by brandonlangley** [comment]
> We need to stop making tests that are predicated on our syntax never changing. @brandonlangley it looks like you created the failing MOOSE server test which perhaps is relying on the number of `MooseMesh` parameters to be exactly equal to 47? My quick perusal and conclusion could be wrong

@lindsayad -

You are correct, my apologies about that.

In January, @GiudGiud fixed a different test checking the top level blocks allowing it to grow with this change: https://github.com/idaholab/moose/commit/70297ec2aa277dfb7693992642f8a445818b780c

I failed to update the test checking these parameters at that time, but I will apply a fix similar to his as soon as possible.

Sorry about the headache.

**Comment by lindsayad** [comment]
No worries. It didn't take me very long to figure out to patch the test

**Review by lindsayad — APPROVED** [review]

File: unit/src/MooseServerTest.C [hunk 9]
```diff
@@ -1045,7 +1045,7 @@ TEST_F(MooseServerTest, CompletionMeshDefaultedType)
 
   EXPECT_EQ(request_id, response_id);
 
-  EXPECT_EQ(47u, completions_array.size());
+  EXPECT_EQ(48u, completions_array.size());
```
Comment (GiudGiud) [inline unit/src/MooseServerTest.C:1048]: this long term fix lol

**Review by GiudGiud — APPROVED** [review]

**Pull request merged**